contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2308,3 +2312,278 @@ get_db_owner_role_name(const char *dbname)
+
+	// get the description of the table
+	bbf_view_def_dsc = RelationGetDescr(bbf_view_def_rel);
+
+	// serach for the row for update => build the key
+	ScanKeyInit(&key[0],
+				Anum_bbf_view_def_object_name,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(stmt->relation->relname));
+	// remove db_name from schema_name for search in bbf_view_def_schemaname
+	curr_schemaname += (strlen(cur_db) + 1);
+	ScanKeyInit(&key[1],
+				Anum_bbf_view_def_schema_name,
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(curr_schemaname));
+
+	// scan
+	tblscan = table_beginscan_catalog(bbf_view_def_rel, 2, key);
+	
+	// get the scan result -> original tuple
+	usertuple = heap_getnext(tblscan, ForwardScanDirection);
+
+	if (!HeapTupleIsValid(usertuple)) {
+		table_endscan(tblscan);
+		table_close(bbf_view_def_rel, RowExclusiveLock);
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				errmsg("Cannot find the view_name \"%s\", because it does not exist or you do not have permission.", stmt->subname)));
+	}
+	
+	// create new tuple to substitute
+	new_record_view_def[Anum_bbf_view_def_object_name - 1] = CStringGetTextDatum(stmt->newname);
+	new_record_repl_view_def[Anum_bbf_view_def_object_name - 1] = true;
+
+	new_tuple = heap_modify_tuple(usertuple,
+								  bbf_view_def_dsc,
+								  new_record_view_def,
+								  new_record_nulls_view_def,
+								  new_record_repl_view_def);
+
+	CatalogTupleUpdate(bbf_view_def_rel, &new_tuple->t_self, new_tuple);
+
+	heap_freetuple(new_tuple);
+
+	table_endscan(tblscan);
+	table_close(bbf_view_def_rel, RowExclusiveLock);
+}
+
+static void
+rename_proc_update_bbf_catalog(RenameStmt *stmt)
+{
+	// update the 'funcname', 'orig_name', 'funcsignature' in 'babelfish_function_ext'
+	Relation		bbf_proc_ext_rel;
+	TupleDesc		bbf_proc_ext_dsc;
+	ScanKeyData		key[2];
+	HeapTuple		usertuple;
+	HeapTuple		new_tuple;
+	TableScanDesc		tblscan;
+	Datum			new_record_proc_ext[BBF_FUNCTION_EXT_NUM_COLS];
+	bool			new_record_nulls_proc_ext[BBF_FUNCTION_EXT_NUM_COLS];
+	bool			new_record_repl_proc_ext[BBF_FUNCTION_EXT_NUM_COLS];
+	NameData		*objname_data;
+	NameData		*schemaname_data;
+	bool			is_null;
+	char			*funcsign, *new_funcsign;
+	Datum			funcsign_datum;
+	Node			*schema;
+	char			*schemaname;
+	ObjectWithArgs	*objwargs = (ObjectWithArgs *)stmt->object;
+
+	// build the tuple to insert
+	MemSet(new_record_proc_ext, 0, sizeof(new_record_proc_ext));
+	MemSet(new_record_nulls_proc_ext, false, sizeof(new_record_nulls_proc_ext));
+	MemSet(new_record_repl_proc_ext, false, sizeof(new_record_repl_proc_ext));
+
+	// open the catalog table
+	bbf_proc_ext_rel = table_open(get_bbf_function_ext_oid(), RowExclusiveLock);
+
+	// get the description of the table
+	bbf_proc_ext_dsc = RelationGetDescr(bbf_proc_ext_rel);
+
+	// search for the row for update => build the key
+	objname_data = (NameData *) palloc0(NAMEDATALEN);
+	snprintf(objname_data->data, NAMEDATALEN, "%s", stmt->subname);
+	ScanKeyInit(&key[0],
```

Comment:
function name alone is not unique. 
In postgresql's case, nspnanme + function name + funcsignature define a unique function or procedure